File: app/src/main/java/com/termux/utils/ControlsUI.java
package com.termux.utils;

import android.app.Activity;
import android.graphics.RenderEffect;
import android.graphics.Shader;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextClock;

import com.termux.NyxActivity;
import com.termux.R;
import com.termux.data.ConfigManager;
import com.termux.terminal.SessionManager;
import com.termux.terminal.TerminalSession;

public final class ControlsUI extends Activity {

    private ViewGroup consoleParent;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setContentView(R.layout.control_ui);
        final int[] child = {-1, -1};

        this.consoleParent = (ViewGroup) NyxActivity.console.getParent();
        final View sessionView = this.findViewById(R.id.session_view);
        final var childCount = this.consoleParent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            if (this.consoleParent.getChildAt(i) instanceof Extrakeys) child[0] = i;
            if (this.consoleParent.getChildAt(i) instanceof FrameLayout) child[1] = i;
        }
        this.setupButton(R.id.new_session, true, v -> {
            SessionManager.addNewSession(false);
            sessionView.requestLayout();
        });
        this.findViewById(R.id.new_session).setOnLongClickListener(v -> {
            SessionManager.addNewSession(true);
            sessionView.requestLayout();
            return true;
        });
        this.setupButton(R.id.plus, true, b -> NyxActivity.console.changeFontSize(true));
        this.setupButton(R.id.minus, true, b -> NyxActivity.console.changeFontSize(false));
        this.setupButton(R.id.extrakeys, -1 != child[0], v -> {
            if (-1 == child[0]) {
                this.consoleParent.addView(new Extrakeys());
                child[0] = this.consoleParent.getChildCount() - 1;
            } else {
                this.consoleParent.removeViewAt(child[0]);
                child[0] = -1;
            }
            ((Button) v).toogle();
        });

        this.setupButton(R.id.win, -1 != child[1], v -> {
            if (-1 == child[1]) {
                this.consoleParent.addView(new WM(this));
                child[1] = this.consoleParent.getChildCount() - 1;
            } else {
                this.consoleParent.removeViewAt(child[1]);
                child[1] = -1;
            }
            ((Button) v).toogle();
        });

        this.setupButton(R.id.close, true, v -> {
            for (final TerminalSession session : SessionManager.sessions) {
                SessionManager.removeFinishedSession(session);
            }
            this.finish();
        });

        final TextClock textClock = this.findViewById(R.id.clock);
        textClock.setTypeface(ConfigManager.typeface);
        textClock.setTextColor(Theme.primary);
        if (Build.VERSION_CODES.S <= Build.VERSION.SDK_INT)
            this.consoleParent.setRenderEffect(RenderEffect.createBlurEffect(6, 5, Shader.TileMode.CLAMP));
    }

    private void setupButton(final int id, final boolean enabled, final View.OnClickListener onClick) {
        final Button button = this.findViewById(id);
        button.setCheck(enabled);
        button.setOnClickListener(onClick);
    }

    @Override
    protected void onDestroy() {
        if (Build.VERSION_CODES.S <= Build.VERSION.SDK_INT)
            this.consoleParent.setRenderEffect(null);
        super.onDestroy();
    }
}

File: app/src/main/java/com/termux/utils/ControlsUI.kt
package com.termux.utils

import android.app.Activity
import android.graphics.RenderEffect
import android.graphics.Shader
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.termux.R
import com.termux.data.ConfigManager
import com.termux.data.console
import com.termux.terminal.SessionManager.addNewSession
import com.termux.terminal.SessionManager.removeFinishedSession
import com.termux.terminal.SessionManager.sessions

class ControlsUI : Activity() {
    private lateinit var console_parent : ViewGroup
    private lateinit var sessionView : View
    override fun onCreate(savedInstanceState : Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.control_ui)
        var extrakeys = -1
        var winman = -1
        console_parent = console.parent as ViewGroup
        sessionView = findViewById(R.id.session_view)

        for (i in 0 until console_parent.childCount) {
            when (console_parent.getChildAt(i)) {
                is Extrakeys -> extrakeys = i
                is WindowManager -> winman = i
            }
        }
        fV(R.id.scroll).requestFocus()
        setupButton(R.id.new_session) {
            addnewSes(false)
        }
        setupButton(R.id.failsafe, false) {
            addnewSes(true)
        }
        setupButton(R.id.extrakeys, extrakeys != -1) {
            extrakeys = if (extrakeys != -1) {
                console_parent.removeViewAt(extrakeys)
                -1
            } else {
                console_parent.addView(Extrakeys())
                console_parent.childCount - 1
            }
            it.toogle()
        }
        setupButton(R.id.win, winman != -1) {
            winman = if (winman != -1) {
                console_parent.removeViewAt(winman)
                -1
            } else {
                console_parent.addView(WindowManager())
                console_parent.childCount - 1
            }
            it.toogle()
        }
        setupButton(R.id.close) {
            sessions.forEach {
                removeFinishedSession(it)
            }
            finish()
        }
        findViewById<TextView>(R.id.clock).typeface = ConfigManager.typeface
        console_parent.setRenderEffect(RenderEffect.createBlurEffect(5f, 5f, Shader.TileMode.CLAMP))
    }

    private fun addnewSes(isFailSafe : Boolean) {
        addNewSession(isFailSafe)
        sessionView.requestLayout()
    }

    private fun setupButton(id : Int, enabled : Boolean = true, onClick : (Button) -> Unit) {
        findViewById<Button>(id).apply {
            setCheck(enabled)
            setOnClickListener { onClick(it as Button) }
        }
    }

    private fun fV(id : Int) = findViewById<View>(id)
    override fun onDestroy() {
        console_parent.setRenderEffect(null)
        super.onDestroy()
    }
}
